spyder/plugins/editor/widgets/completion.py

```diff
@@ -0,0 +1,401 @@
+# (see spyder/__init__.py for details)
+
+"""Completion widget class"""
+
+# Standard library imports
+import sys
+
+# Third psrty imports
+from qtpy.QtCore import QPoint, Qt, Signal, Slot
+from qtpy.QtGui import QFontMetrics
+from qtpy.QtWidgets import (QAbstractItemView, QApplication, QListWidget,
+                            QListWidgetItem, QToolTip)
+
+# Local imports
+from spyder.utils import icon_manager as ima
+from spyder.plugins.completion.languageserver import CompletionItemKind
+from spyder.py3compat import to_text_string
+from spyder.widgets.helperwidgets import HTMLDelegate
+
+
+COMPLETION_ITEM_TEMPLATE = """
+<table width="{width}" height="{height}">
+    <tr>
+        <td valign="middle" style="color:{color}">
+            {completion}
+        </td>
+
+        <td valign="middle" align="right" float="right" style="color:{color}">
+            {type}
+        </td>
+    </tr>
+</table>
+"""
+DEFAULT_COMPLETION_ITEM_HEIGHT = 15
+DEFAULT_COMPLETION_ITEM_WIDTH = 250
+
+
+class CompletionWidget(QListWidget):
+    """Completion list widget"""
+
+    sig_show_completions = Signal(object)
+    sig_completion_hint = Signal(str, str, QPoint)
+
+    def __init__(self, parent, ancestor):
+        QListWidget.__init__(self, ancestor)
+        self.setWindowFlags(Qt.SubWindow | Qt.FramelessWindowHint)
+        self.textedit = parent
+        self.hide()
+        self.itemActivated.connect(self.item_selected)
+        self.currentRowChanged.connect(self.row_changed)
+        self.is_internal_console = False
+        self.completion_list = None
+        self.completion_position = None
+        self.automatic = False
+
+        # Setup item rendering
+        self.setItemDelegate(HTMLDelegate(self, margin=3))
+        self.setMinimumWidth(DEFAULT_COMPLETION_ITEM_WIDTH)
+
+        # Initial item height and width
+        fm = QFontMetrics(self.textedit.font())
+        self.item_height = fm.height()
+        self.item_width = self.width()
+
+    def setup_appearance(self, size, font):
+        self.resize(*size)
+        self.setFont(font)
+
+    def is_empty(self):
+        """Check if widget is empty."""
+        if self.count() == 0:
+            return True
+        return False
+
+    def show_list(self, completion_list, position, automatic):
+        """Show list corresponding to position."""
+
+        if not completion_list:
+            self.hide()
+            return
+
+        self.automatic = automatic
+
+        if position is None:
+            # Somehow the position was not saved.
+            # Hope that the current position is still valid
+            self.completion_position = self.textedit.textCursor().position()
+
+        elif self.textedit.textCursor().position() < position:
+            # hide the text as we moved away from the position
+            self.hide()
+            return
+
+        else:
+            self.completion_position = position
+
+        # Completions are handled differently for the Internal
```

Comment:
Should we update the internal console (on a separate future PR) so that we have a single  API @ccordoba12 ?

Reply:
Sure, that would be good.

Reply:
@dalthviz please open an issue to make the internal shell have the same API